glib/src/value.rs

```diff
@@ -56,31 +51,264 @@ use crate::gstring::GString;
 use crate::translate::*;
 use crate::types::{StaticType, Type};
 
-/// An error returned from the [`get`](struct.Value.html#method.get)
-/// or [`get_some`](struct.Value.html#method.get_some) functions on a [`Value`](struct.Value.html)
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct GetError {
-    pub actual: Type,
-    pub requested: Type,
+/// A type that can be stored in `Value`s.
+pub trait ValueType: ToValue + for<'r> FromValue<'r> {
+    /// Type to get the `Type` from.
+    ///
+    /// This exists only for handling optional types.
+    // FIXME: Should default to Self once associated type defaults are stabilized
+    // https://github.com/rust-lang/rust/issues/29661
+    type Type: StaticType;
+}
+
+/// A type that can be stored in `Value`s and is optional.
+///
+/// These are types were storing an `Option` is valid. Examples are `String` and all object types.
+pub trait ValueTypeOptional:
+    ValueType
+    + ToValueOptional
+    + for<'r> FromValue<'r, Error = ValueTypeMismatchOrNoneError<Self>>
+    + StaticType
+{
+}
+
+impl<'r, T: ValueTypeOptional + StaticType> ValueType for Option<T>
+where
+    Self: for<'s> FromValue<'s, Error = ValueTypeMismatchOrNoneError<T>>,
+{
+    type Type = T::Type;
+}
+
+/// Trait for errors returned from `Value::get`.
+pub unsafe trait ValueTypeChecker: std::error::Error + Sized {
```

Comment:
Will become `pub unsafe trait ValueTypeChecker: std::error::Error + Sized + Send + Sync + 'static` {` so that this can be used painlessly

Reply:
One problem here is that `ValueTypeMismatch<T>` is only `'static` if `T: 'static` but that's not really what we want here (e.g. `&str`). It doesn't matter that the `T` is only used in a `PhantomData` here.

Reply:
Can you give an example of the pain if it wasn't static?

Reply:
For using it as `std::error::Error::source()` return value (the `gst::Structure` getter error)

Reply:
Generally, non-`'static` errors are not great for use in combination with other code.

Reply:
Which is of course also a problem for the structure getter error because it borrows the field name string

Reply:
Ok, so my plan now is

```rust
pub trait ValueTypeChecker {
  type Error = ...;
  fn check(&Value) -> Result<(), Self::Error>;
}

pub struct ValueTypeMismatchError { actual: Type, requested: Type }

pub struct GenericValueTypeChecker<T>;

impl<T: Type> ValueTypeChecker for GenericValueTypeChecker<T> {
  type Error = ValueTypeMismatchError;
  fn check...
}

pub trait FromValue {
  type Checker: ValueTypeChecker;
  ...
}

etc.
```

That way the error types and everything are `'static` etc.

Also, for GStreamer's structure getter: the field names are actually `GQuark`s so can be just `&'static str`.